src/main/java/htsjdk/samtools/util/IOUtil.java

```diff
@@ -346,34 +347,12 @@ public static Path getDefaultTmpDirPath() {
     }
 
     /**
-     * Deletes on exit a path by creating a shutdown hook.
-     */
-    public static void deleteOnExit(final Path path) {
-        // add a shutdown hook to remove the path on exit
-        Runtime.getRuntime().addShutdownHook(new DeletePathThread(path));
-    }
-
-    /**
-     * WARNING: visible for testing. Do not use.
-     *
-     * Class for delete a path, used in a shutdown hook for delete on exit.
+     * Register a {@link Path} for deletion on JVM exit.
      *
-     * @see #deleteOnExit(Path)
+     * @see DeleteOnExitPathHook
      */
-    static final class DeletePathThread extends Thread {
-
-        private final Path path;
-
-        DeletePathThread(Path path) {this.path = path;}
-
-        @Override
-        public void run() {
-            try {
-                Files.deleteIfExists(path);
-            } catch (IOException e) {
-                throw new RuntimeIOException(e);
-            }
-        }
+    public static void deleteOnExit(final Path path) {
+        DeleteOnExitPathHook.add(path);
```

Comment:
Why not call `path.toFile().deleteOnExit()` here?

Reply:
So on HDFS this will throw? Can we special case HDFS by examining `Path.getFileSystem()`?

Do we have cases where we'll be creating temporary files in HDFS or is this a theoretical case?

Reply:
@pshapiro4broad The reason we're adding support for Path is to support file systems that can't be represented as files, i.e. hdfs://, gs://, s3://, etc.   We definitely want to be able to make temp files on hdfs, gcs, etc.  Not so much for production work, it wouldn't make much sense to store temp files on gcs, but it's incredibly use for testing.

Reply:
OK. I think we'd be better off with a more test specific way to delete test data, and avoid using shutdown hooks complete, but that seems unlikely to change at this point.

Reply:
I think that it is a common practice to use `File.deleteOnExit` for some temp files (not only for test data, but also for temporary files written on disk) because it is better to use the shutdown hook as it will run always and thus not keeping the tmp files if there is a crash that does not finalize.

For `Path`, there is not such a method. This commit here is just using the same approach as the `File` version, to be in-line with what it is expected if we change HTSJDK to use only `Path`. For example, this is also used in the sorting collection and using it with a, for example, HDFS/GCS `Path` as a temporary directory requires that this method remove the files on shutdown of the JVM.

Reply:
It's a common practice in this codebase but it's not IMO a good practice, and is not common everywhere. A better system is to make the code define the scope of all resources, so that when a resource goes out of scope, it can be deleted. There is no need to use a finalizer, using a `finally` clause (or its equivalent) is sufficient.

Reply:
@pshapiro4broad The problem with that is that it's a pain and people don't do it when they write tests so they just leave files everywhere.  This is the best way I know that doesn't impose a lot of boilerplate on the test code.